[VGNWB-209] added functionality to delete relation templates

prnvshrn opened the pull request.

Commit:
[VGNWB-209] added close button to delete relation template

Commit:
[VGNWB-209] added close button to delete relation template

Commit:
[VGNWB-209]

Commit:
[VGNWB-209] added functionality to delete relation templates

Commit:
[VGNWB-209] added functionality to delete relation templates

jdamerow commented on annotations/templates/annotations/relationtemplate_list.html:
```diff
@@ -26,16 +26,61 @@
+                        </div>
+                      </div>
+                    </div>
+                    <!-- Delete Template modal section ends here -->
```
I wouldn't create a modal for every single template. That just becomes unmanageable when there are many templates. Rather have one modal for which you set the selected template id when the delete button is clicked.

prnvshrn replied:
Changed the implementation of creating multiple modals to a single modal.

jdamerow commented on annotations/templates/annotations/relationtemplate_list.html:
```diff
@@ -46,9 +91,21 @@
+
+    $('[data-delete-template]').tooltip();
+
+    $("a[id^='openTemplateDeleteModal']").click(function(e) {
```
just give all buttons the same class and then use the class to identify buttons here.

prnvshrn replied:
The buttons now are identified by class.

jdamerow commented on annotations/views/relationtemplate_views.py:
```diff
@@ -208,5 +206,12 @@ def create_from_relationtemplate(request, template_id):
     return JsonResponse(response_data)
+
+
+def delete_relationtemplate(request, template_id):
```
this needs to be access restricted.

prnvshrn replied:
Added "staff_member_required" before <b>delete_relationtemplate</b> method to restrict access.

Commit:
[VGNWB-209] changed delete template modal implementation to a single modal

prnvshrn commented:
I have made the following changes:-
1. Changed multiple modals to a single modal
2. Using class to identify buttons for opening the modal
3. Added 'staff_member_required' to restrict access

jdamerow commented on annotations/views/relationtemplate_views.py:
```diff
@@ -208,5 +206,13 @@ def create_from_relationtemplate(request, template_id):
+def delete_relationtemplate(request, template_id):
+    if request.method == 'POST':
+        RelationTemplate.objects.filter(id=template_id).delete()
+        RelationTemplatePart.objects.filter(id=template_id).delete()
```
this should be done in a transaction:
https://docs.djangoproject.com/en/1.11/topics/db/transactions/#controlling-transactions-explicitly

prnvshrn replied:
Changed implementation of deletion logic to execute in an atomic transaction

Commit:
[VGNWB-209] changed deletion of relation templates into atomic transaction

prnvshrn commented:
Changed implementation of deletion logic to execute in an atomic transaction

jdamerow commented:
something is going wrong here. try this: 
- create 3 templates (two simple subject-predicate-object ones, and the last one nested, not sure if this is important, but that's what I did). 
- then delete the last one (the nested one).
- no look at the first one: -> relation parts are deleted

jdamerow commented:
also, it looks like relationsets have references to relation templates, which means relation templates are deleted that are referenced by relationsets, there will most likely be unintended side effects. A relation template should probably not be allowed to be deleted if it is used in a relation set.

Commit:
[VGNWB-209] added check for not deleting template if data in relationset

Commit:
[VGNWB-209] added check for checking if data exists for relation template

prnvshrn commented:
Added the following changes -
1. Wrong deletion of relation template part was occurring because in the views I was passing <b>id</b> instead of <b>part_of</b> for deletion. So I changed the method to use <b>part_of</b> instead.
2. Added check to see if data is present in relationset for a template before deletion. If data is present, an error is shown and template is not deleted.
